src/snmp_agent.c

```diff
@@ -0,0 +1,1513 @@
+                             &td->size_oid.oid_len)) {
+    ERROR(PLUGIN_NAME ": Failed to parse table size OID (%s)",
+          ci->values[0].value.string);
+    td->size_oid.oid_len = 0;
+    return (-EINVAL);
+  }
+
+  return (0);
+}
+
+static int snmp_agent_config_table_index_oid(table_definition_t *td,
+                                             oconfig_item_t *ci) {
+
+  if (ci->values_num < 1) {
+    WARNING(PLUGIN_NAME ": `IndexOID' is empty.");
+    return (-EINVAL);
+  }
+
+  if (ci->values[0].type != OCONFIG_TYPE_STRING) {
+    WARNING(PLUGIN_NAME ": `IndexOID' needs only string argument.");
+    return (-EINVAL);
+  }
+
+  td->index_oid.oid_len = MAX_OID_LEN;
+
+  if (NULL == snmp_parse_oid(ci->values[0].value.string, td->index_oid.oid,
+                             &td->index_oid.oid_len)) {
+    ERROR(PLUGIN_NAME ": Failed to parse table index OID (%s)",
+          ci->values[0].value.string);
+    td->index_oid.oid_len = 0;
+    return (-EINVAL);
+  }
+
+  return (0);
+}
+
+static int snmp_agent_config_table_data(table_definition_t *td,
+                                        oconfig_item_t *ci) {
+  data_definition_t *dd;
+  int ret = 0;
+
+  assert(ci != NULL);
+
+  dd = calloc(1, sizeof(*dd));
+  if (dd == NULL) {
+    ERROR(PLUGIN_NAME ": Failed to allocate memory for table data definition");
+    return (-ENOMEM);
+  }
+
+  ret = cf_util_get_string(ci, &dd->name);
+  if (ret != 0) {
+    sfree(dd);
+    return (-1);
+  }
+
+  dd->scale = 1.0;
+  dd->shift = 0.0;
+
+  dd->table = td;
+
+  for (int i = 0; i < ci->children_num; i++) {
+    oconfig_item_t *option = ci->children + i;
+
+    if (strcasecmp("Instance", option->key) == 0)
+      ret = cf_util_get_boolean(option, &dd->is_instance);
+    else if (strcasecmp("Plugin", option->key) == 0)
+      ret = cf_util_get_string(option, &dd->plugin);
+    else if (strcasecmp("PluginInstance", option->key) == 0)
+      ret = cf_util_get_string(option, &dd->plugin_instance);
+    else if (strcasecmp("Type", option->key) == 0)
+      ret = cf_util_get_string(option, &dd->type);
+    else if (strcasecmp("TypeInstance", option->key) == 0)
+      ret = cf_util_get_string(option, &dd->type_instance);
+    else if (strcasecmp("Shift", option->key) == 0)
+      ret = cf_util_get_double(option, &dd->shift);
+    else if (strcasecmp("Scale", option->key) == 0)
+      ret = cf_util_get_double(option, &dd->scale);
+    else if (strcasecmp("OIDs", option->key) == 0)
+      ret = snmp_agent_config_data_oids(dd, option);
+    else {
+      WARNING(PLUGIN_NAME ": Option `%s' not allowed here.", option->key);
+      ret = -1;
+    }
+
+    if (ret != 0) {
+      snmp_agent_free_data(&dd);
+      return (-1);
+    }
+  }
+
+  llentry_t *entry = llentry_create(dd->name, dd);
+  if (entry == NULL) {
+    snmp_agent_free_data(&dd);
+    return (-ENOMEM);
+  }
+
+  llist_append(td->columns, entry);
+
+  return (0);
+}
+
+static int snmp_agent_config_data(oconfig_item_t *ci) {
```

Comment:
This duplicates a lot of code from `snmp_agent_config_table_data()`. Could this be simplified?